src/main/java/gregtech/api/unification/stack/MultiItemVariantMap.java

```diff
@@ -0,0 +1,93 @@
+package gregtech.api.unification.stack;
+
+import gregtech.api.GTValues;
+import it.unimi.dsi.fastutil.shorts.Short2ObjectMap;
+import it.unimi.dsi.fastutil.shorts.Short2ObjectOpenHashMap;
+
+import javax.annotation.Nullable;
+
+/**
+ * {@link ItemVariantMap} implementation backed by a hashmap. Each metadata is
+ * treated as separate entry in hashmap, and get/set accesses treat each metadata
+ * as a unique item variant.
+ *
+ * @param <E> type of the elements
+ */
+public final class MultiItemVariantMap<E> implements ItemVariantMap.Mutable<E> {
+
+    @Nullable
+    private Short2ObjectOpenHashMap<E> itemDamageEntries;
+    @Nullable
+    private E wildcardEntry;
+
+    @Override
+    public boolean hasNonWildcardEntry() {
+        return this.itemDamageEntries != null && !this.itemDamageEntries.isEmpty();
+    }
+
+    @Override
+    public boolean hasEntry(short meta) {
+        if (meta == GTValues.W) {
+            return this.wildcardEntry != null;
+        } else {
+            return this.itemDamageEntries != null && this.itemDamageEntries.containsKey(meta);
+        }
+    }
+
+    @Nullable
+    @Override
+    public E getEntry(short meta) {
+        if (meta == GTValues.W) {
+            return this.wildcardEntry;
+        } else if (this.itemDamageEntries != null) {
+            return this.itemDamageEntries.get(meta);
+        } else {
+            return null;
+        }
+    }
+
+    @Nullable
+    @Override
+    public E setEntry(short meta, @Nullable E e) {
+        if (meta == GTValues.W) {
+            E cache = this.wildcardEntry;
+            this.wildcardEntry = e;
+            return cache;
+        } else if (e != null) {
+            if (this.itemDamageEntries == null) {
+                this.itemDamageEntries = new Short2ObjectOpenHashMap<>();
+            }
+            return this.itemDamageEntries.put(meta, e);
+        } else {
+            if (this.itemDamageEntries != null) {
+                return this.itemDamageEntries.remove(meta);
+            } else {
+                return null;
+            }
+        }
+    }
+
+    @Override
+    public void clear() {
+        this.itemDamageEntries = null;
```

Comment:
Would it be preferred to call `Map#clear` on `itemDamageEntries` (if it's nonnull), instead of setting it to null?

Reply:
I don't think either of them will affect the performance in any capacity (because the `clear` impl in question is never called xd). Clearing entry might be better on some use cases, but I can't think of what that use case would be.